### DISCUSSION TITLE
Use pre-installed ChromeDriver with Python on Ubuntu-latest

### DISCUSSION THREAD

**Discussion by yusancky** [community/community#44279] [Actions] [answered]
Previously, I used to install a separate browser driver when using Selenium. But in fact, [GitHub Actions runner already comes with ChromeDriver pre-installed](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md#browsers-and-drivers), so I should be able to get away without installing it separately. So, I modified the code. 

```python
from os import environ
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service

chrome_service = Service(environ['CHROMEWEBDRIVER'])
chrome_options = Options()
for option in ['--headless','--disable-gpu','--window-size=1920,1200','--ignore-certificate-errors','--disable-extensions','--no-sandbox','--disable-dev-shm-usage']:
    chrome_options.add_argument(option)
driver = webdriver.Chrome(service = chrome_service,options = chrome_options)
```

However, I triggered an Actions manually and it came up with the following error message. 

```powershell
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/selenium/webdriver/common/service.py", line 71, in start
    self.process = subprocess.Popen(cmd, env=self.env,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/subprocess.py", line 1024, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/subprocess.py", line 1901, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: '/usr/local/share/chrome_driver'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/repo-name/repo-name/file-name.py", line 15, in <module>
    driver = webdriver.Chrome(service = chrome_service,options = chrome_options)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/selenium/webdriver/chrome/webdriver.py", line 69, in __init__
    super().__init__(DesiredCapabilities.CHROME['browserName'], "goog",
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/selenium/webdriver/chromium/webdriver.py", line 89, in __init__
    self.service.start()
  File "/opt/hostedtoolcache/Python/3.11.1/x64/lib/python3.11/site-packages/selenium/webdriver/common/service.py", line 86, in start
    raise WebDriverException(
selenium.common.exceptions.WebDriverException: Message: 'chrome_driver' executable may have wrong permissions. Please see https://chromedriver.chromium.org/home
```

#26275 seems to raise similar issues, but I don't know how to use it or if it applies here. 

Some websites say that running `sudo chmod +x /usr/local/share/chrome_driver` before running Python code will fix it, but that didn't work for me.

**Comment by yusancky** [comment, marked as answer]
Just now, I thought that if I didn't have enough permissions to access a folder, I could copy one to use with the `cp` command. It turned out that this command produced an error - it suggested that `/usr/local/share/chrome_driver` was a folder. 

I instantly understood, and added a new job, which looked like this: 
```yml
- name: Tree
  run: |
    cd /usr/local/share/chrome_driver
    tree
```

I got the following output: 
```powershell
.
└─ chromedriver

0 directories, 1 file
```

Yeah! It's clear that I should use the file called `chromedriver`, instead the folder named `chrome_driver`. **So just change `/usr/local/share/chrome_driver` to `/usr/local/share/chrome_driver/chromedriver` can solve the problem.** 

This is actually a bit of an oversight on my part: the documentation gives the value of the environment variable, which should obviously be a folder, but I didn't think of it.

**Reply by wonderingwombat** [reply]
Thanks for this, I wasted way too much time trying to fix this until I read your comment.

**Reply by mckinlde** [reply]
Thanks2: I had a different issue but ended up using this doc to fix it.  If anyone like me ends up here looking for a way to use selenium with pre-installed browser *and* driver binaries, here's a boilerplate repo:
https://github.com/mckinlde/static_binary_selenium_webdriver_chrome_121

**Comment by m7mad120059** [comment]
الووو **[]()**
